### ISSUE TITLE
About loss function

### ISSUE THREAD

**Issue by wudongming97** [DerrickWang005/CRIS.pytorch#5]
Hi, I found that the loss used in this repo is a cross-entropy loss between prediction and mask.

`loss = F.binary_cross_entropy_with_logits(pred, mask)`

But the loss mentioned in the paper is a contrastive loss between visual and textual features.

**Comment by Deepayan137** [comment]
Hello! I wrote the contrastive learning part by following the instructions in the paper. However, when training the model only with the contrastive loss, the training IOU doesn't seem to improve. Below, I am attaching the code snippet and the training IOU and precision curves. The training is done only for 1 epoch. The brown plots are for cross-entropy loss while the blue plots are for contrastive loss.  I would be grateful if you could let me know what I am doing wrong and also if the contrastive loss is supposed to be used in addition to cross-entropy loss.
Thanks
```
def forward(self, x, word, mask):
        x = self.vis(x)
        B, C, H, W = x.size()
        word = self.txt(word)
        x = x.permute(0, 2, 3, 1)
        out = torch.einsum('nhwc,nc->nhw', x, word).unsqueeze(1)
        out = torch.sigmoid(out) #sigmoid of zt dot zv
        loss = torch.zeros((H, W)).cuda()
        pos_count, neg_count = 0, 0
        for i in range(word.size(0)):
            zt = word[i]
            zt = zt.unsqueeze(0)
            for j in range(x.size(0)):
                zv = x[j]
                zv = zv.reshape(self.in_dim, -1)
                prod = torch.mm(zt, zv).squeeze()
                prod = prod.reshape(H, W)
                if i == j:
                    pos = - torch.log(F.sigmoid(prod))
                    loss += pos
                    pos_count += 1
                else:
                    neg = - torch.log(1 - F.sigmoid(prod))
                    loss += neg
                    neg_count += 1
        total = pos_count + neg_count
        loss = torch.mean(loss)
        if out.shape[-2:] != mask.shape[-2:]:
            mask = F.interpolate(mask, out.shape[-2:],
                mode='nearest').detach()
        return out, loss/total, mask
```
![W B Chart 09_07_2022, 10_18_53](https://user-images.githubusercontent.com/21141806/178092057-453c7050-2c4a-4fbb-99c2-5f1dccd69ba2.png)
![W B Chart 09_07_2022, 10_18_40](https://user-images.githubusercontent.com/21141806/178092073-d7574f3c-3b83-4ef9-b167-7c022ba8921b.png)

**Comment by DerrickWang005** [comment]
please follow our implementation. https://github.com/DerrickWang005/CRIS.pytorch/blob/0df39f073acfb9e6e17d83536a916548905ecfc3/model/layers.py#L47-L84

**Comment by Deepayan137** [comment]
Hello Derrick,

I had seen this implementation. In your paper, you have mentioned equations 9 and 10 as the contrastive loss between pixel embeddings and the text features. I am not able to understand, how it is taken care of in your above code snippet?

**Comment by tiger990111** [comment]
I have the same query. Can the authors please clarify?

**Comment by FabianRitter** [comment]
No follow up? looks like supervised learning on the code. I assume something is missing in the code.

**Comment by Starboy-at-earth** [comment]
@DerrickWang005 Could you please realse the code snippet of contrastive learning loss?

**Comment by wildwolff** [comment]
After reading the code, I think the authenticity of this paper has yet to be verified

**Comment by Fake10086** [comment]
you may take a deeper look at codes mentioned by the author above, and you'll find that conv2d actually acts like element wise product between text and image which can be considered as equation 9&10.

**Comment by lyu-yx** [comment]
I have the same question, could the authors release the latest version of code?

**Comment by DerrickWang005** [comment]
I think this article can answer your question to some extent. @lyu-yx 
https://arxiv.org/pdf/2303.15343.pdf

**Comment by ccccchenllll** [comment]
I have the same question. I couldn't find the code about contrastive loss.

**Comment by Shaosifan** [comment]
> I have the same question. I couldn't find the code about contrastive loss.

Me too...

**Comment by Kimsure** [comment]
I prefer that Eq. 9 and Eq. 10 actually represent a BCE loss, and agree that this is actually not the standard contrastive learning. @clownrat6
